s3_resource/src/s3_operations.cpp

```diff
@@ -171,195 +182,226 @@ namespace irods_s3 {
                                                       bool query_metadata = true) -> irods::error
     {
         using logger_config = irods::experimental::log::logger_config<s3_plugin_logging_category>;
+        using named_shared_memory_object =
+            irods::experimental::interprocess::shared_memory::named_shared_memory_object
+            <multipart_shared_data>;
         std::uint64_t thread_id = std::hash<std::thread::id>{}(std::this_thread::get_id());
 
-        // ********* DEBUG - print L1desc for all
-        if (logger_config::get_level() == log::level::debug ||  logger_config::get_level() == log::level::trace) {
+        irods::file_object_ptr file_obj = boost::dynamic_pointer_cast<irods::file_object>(_ctx.fco());
 
-            logger::debug("{}:{} ({}) [[{}]] ------------- L1desc ---------------",
-                    __FILE__, __LINE__, __FUNCTION__, thread_id);
-            for (int i = 0; i < NUM_L1_DESC; ++i) {
-                if (L1desc[i].inuseFlag && L1desc[i].dataObjInp && L1desc[i].dataObjInfo) {
-                   int thread_count = L1desc[i].dataObjInp->numThreads;
-                   int oprType = L1desc[i].dataObjInp->oprType;
-                   std::int64_t data_size = L1desc[i].dataSize;
-                   logger::debug("{}:{} ({}) [[{}]] [{}][objPath={}][filePath={}][oprType={}]"
-                           "[requested_number_of_threads={}][dataSize={}][dataObjInfo->dataSize={}][srcL1descInx={}]",
-                           __FILE__, __LINE__, __FUNCTION__, thread_id, i, L1desc[i].dataObjInp->objPath,
-                           L1desc[i].dataObjInfo->filePath, oprType, thread_count, data_size,
-                           L1desc[i].dataObjInfo->dataSize, L1desc[i].srcL1descInx);
+        // Open shared memory and see if we know the number of threads from another thread
+        std::string shmem_key = SHARED_MEMORY_KEY_PREFIX +
+            std::to_string(std::hash<std::string>{}(get_resource_name(_ctx.prop_map()) + file_obj->logical_path()));
+        logger::trace("{}:{} ({}) [[{}]] shmem_key={} hashed_string={}", __FILE__, __LINE__, __FUNCTION__, thread_id, shmem_key, get_resource_name(_ctx.prop_map()) + file_obj->logical_path() );
+
+        named_shared_memory_object shm_obj{shmem_key,
+            DEFAULT_SHARED_MEMORY_TIMEOUT_IN_SECONDS,
+            SHMEM_SIZE};
+
+        // wrapping this in an atomic_exec so only one thread/process for a specific data object is executed at a time
+        std::string func(__FUNCTION__);
```

Comment:
`__FUNCTION__` is not standard c++. Prefer `__func__`.

Reply:
done